feat: add require stack & surface Yarn PnP

MarcoScabbiolo opened the pull request:
<!-- Please remember to update CHANGELOG.md in the root of the project if you have not done so. -->

## Summary

- Add a require stack printed relative to `<rootDir>` for convenience and linking when a `MODULE_NOT_FOUND` error is thrown by the resolver.
- Throw errors thrown by Yarn PnP as-is. Jest was swallowing them and failing later attempting to resolve the module.

Solves #9594 

## Test plan

E2E tests included in `pnp` and `resolve`

SimenB commented on packages/jest-runtime/src/helpers.ts:
```diff
@@ -53,3 +54,12 @@ export const findSiblingsWithFileExtension = (
+): string => `
+
+Require stack:
+  ${requireStack.map(p => p.replace(`${rootDir}${path.sep}`, '')).join('\n  ')}
```
we have a helper called `replaceRootDirInPath` in `jest-config`, please use it

SimenB replied:
also, please pass the path through `slash` to normalize path separators to unix style (we do this for most paths shown to the user already)

SimenB commented on packages/jest-runtime/src/index.ts:
```diff
@@ -158,6 +159,7 @@ class Runtime {
     this._sourceMapRegistry = Object.create(null);
     this._fileTransforms = new Map();
     this._virtualMocks = Object.create(null);
+    this._requireStack = [];
```
I don't think the stack should be per runtime, but per module.

Looking at node's implementation, they construct the array at runtime when not finding a module by walking up the `module.parent` property - I think we should do the same.

https://github.com/nodejs/node/blob/ecfb7b0988da93b1b140f7fbe877824296697364/lib/internal/modules/cjs/loader.js#L1020-L1034

SimenB commented on packages/jest-runtime/src/index.ts:
```diff
@@ -1286,6 +1287,28 @@ class Runtime {
     ];
   }
+
+  private handleExecutionError(e: Error, module: InitialModule): void {
```
```suggestion
  private handleExecutionError(e: Error, module: InitialModule): never {
```

SimenB commented on packages/jest-runtime/src/RuntimeModuleNotFoundError.ts:
```diff
@@ -0,0 +1,42 @@
+  public requireStack?: Array<Config.Path>;
+  public stackTraceBuilt?: boolean;
+
+  public static buildMessage(
```
should we just add this to `ModuleNotFoundError` directly instead of extending it?

SimenB commented on packages/jest-runtime/src/RuntimeModuleNotFoundError.ts:
```diff
@@ -0,0 +1,42 @@
+
+import * as path from 'path';
+import {ModuleNotFoundError} from 'jest-resolve';
+import {Config} from '@jest/types';
```
```suggestion
import type {Config} from '@jest/types';
```

You might need to rebase on master for this first

SimenB commented on e2e/__tests__/resolveNoFileExtensions.test.ts:
```diff
@@ -20,6 +20,7 @@ test('show error message with matching files', () => {
   const {rest} = extractSummary(stderr);
 
   expect(exitCode).toBe(1);
+
```
```suggestion
```

just to reduce the diff

MarcoScabbiolo commented on packages/jest-runtime/src/index.ts:
```diff
@@ -32,6 +32,7 @@ import {
 import type {V8CoverageResult} from '@jest/test-result';
 import {CoverageInstrumenter, V8Coverage} from 'collect-v8-coverage';
 import * as fs from 'graceful-fs';
+import type ModuleNotFoundError from 'jest-resolve/build/ModuleNotFoundError';
```
@SimenB I am not sure about this import, but since `jest-resolve`'s `index.ts` exports using `export = Resolver`, changing that `export default` to then support `export type ...` would break how other modules use `jest-resolve`.

This is what VS Code and Typescript suggested, it is not clean by any means but I don't see any other way of importing the type (not the actual instance of the class, because it is exporte by reference in `static Resolver.ModuleNotFoundError = ModuleNotFoundError`

SimenB replied:
we can export a

```ts
function isModuleNotFoundError(error: Error): error is ModuleNotFoundError {
  return (error as any).code === 'E_MODULE_NOT_FOUND';
}
```

from `jest-resolve` and call that in `jest-runtime` instead of casting and the `.code` check there. Then you don't need to import it at all

MarcoScabbiolo replied:
Good approach, its called `tryCastModuleNotFoundError`

MarcoScabbiolo commented:
It should be all good now.

I've fixed another bug, when there is some kind of nesting of required modules (`index.js` > `module1.js` > `module2.js`). The "However Jest was able to find `moduleName`" error wiill always show up since the error will bubble up from `module2` to `module1` and when it looks for a sibling module with similar file extension, it will find the module itself (`module1.js`) and show the hint for a file that actually exists and is there.

I've fixed it by adding a `moduleName` property on `ModuleNotFoundError` to hold a reference to the actual module that could not be found, so when the runtime attempts to find a sibling file with a different extension looks only for a file that was not found. The test `should throw module not found error if the module has dependencies that cannot be found` at `e2e/resolve/__tests__/resolve.test.js` covers this scenario.

There should not be any breaking changes AFAIK. In case a `ModuleNotFoundError` is thrown by a resolver that is not the default one the runtime will still be able tu duck type it, and use the previous behaviour in case the resolver did not specify the module that failed to require (`moduleName` property).
